shared-module/capsuleio/__init__.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Jonah Yolles-Murphy (TG-Techie)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/capsuleio/__init__.h"
+#include "py/objstr.h"
+#include <string.h>
+
+
+// the time capsule memory reservation
+capsuleio_capsule_t capsuleio_capsule;
```

Comment:
I do not think think this is a reasonable thing to do. Have you considered such boards as the QT PY? This will take 1/32 of its total memory. Probably many folks are running those right on the edge and 1kb would push them over the edge and unable to update.

What does the data sheet say on typical circuitpython boards w.r.t. program/erase cycles on the embedded flash, and what’s the typical cycle count on express board flash? I imagine if you take into account whatever load leveling is there you’ll find that it’s expected to live decades or more with a 1kb text file written occasionally. (I.e., is this solving a real problem?)

Maybe instead you want to learn more about the garbage collector & memory allocator and mark a capsuled tree as uncollectible? In this way the memory cost would be dynamic and opt-in rather than a high-water mark for the largest allocation somebody changes the capsuleio_amount_bytes const to.

Reply:
`alarm.sleep_memory` and `microcontroller.nvm` both provide persistent storage with a simple bytearray-style of storage. Rather than pass a string, arbitrary bytes is more flexible, though a bit less convenient. Anything can be converted to bytes. The API of those modules would be what to follow. This would be called `supervisor.persistent_memory` or something like that. (We need to think of a good name.)

@WarriorOfWire I have seen use cases of needing persistent storage and restarting something every few seconds. For some chips I calculated this would reach the specified flash lifetime after a few weeks.

Rather than reserving a chunk of RAM permanently, the data in question can be copied out of the heap to a safe place when the VM exits, and then put back after the restart. The "safe place" might be on the stack, or just someplace that's untouched while the VM is not running. It might even work to leave it in its old heap location. As soon as the heap is re-created, a new heap object of the right size can be created and the old data copied there.

Reply:
## API
> `alarm.sleep_memory` and `microcontroller.nvm` both provide persistent storage with a simple bytearray-style of storage

@dhalbert That sounds like a much better interface, especially since that mirrors the literal, physical underlying hardware. I really like your suggestion of putting it with the supervisor namespace as the "time capsule"\* should share the same lifetime.  

<br>

@WarriorOfWire: Agreed 1k is not a reasonable amount of memory for non-full builds of cp. The ..._CAPSULE_AMOUNT_BYTES should definitely change for the size of the build (maybe m0 = 128|256, m4/nrf 512|1024, imx=even more?). 
It's also not uncommon for  'niceties' like gamepad, json, or ulab to be excluded from small builds, maybe only full builds of cp should include this? 


## Implementation

> Maybe instead you want to learn more about the garbage collector & memory allocator and mark a capsuled tree as uncollectible?

Opt-in does sound a lot better than consistent memory hit and at the same time I'd be wary to mark an arbitrary tree of python objects as 'do not collect'? Though the `supervisor.time_capsule*` interface would not have that option.


> Rather than reserving a chunk of RAM permanently, the data in question can be copied out of the heap to a safe place when the VM exits, and then put back after the restart. 

Could you elaborate, please? What is the advantage of copying in and out of the safe place vs storing directly to the safe place? 

## separate thought
A side note on naming, 'persistent' can be another name for 'non-volatile' in the memory hierarchy. IIRC previous conversations have discussed calling it "scratchpad" but that may not be clear.

An additional difference is on disk and `.nvm` will persist between resets, not just reloads. 
The guarantee that the "time capsule"\* will always init to all zeros on reset. I think there is some safety there, though that was not a motivating reason to implement this.

\* whatever the final name becomes

Reply:
> Could you elaborate, please? What is the advantage of copying in and out of the safe place vs storing directly to the safe place?

It doesn't have to be pre-allocated and use up precious space in RAM if not used.

I have a somewhat better API idea:

Don't provide a fixed-size area but to instead just do something like `supervisor.persisted_bytes = some bytearray or bytes`. So it could be as large or small as you want. When the VM exits, the bytes are either left where they are (as described above) or copied to someplace temporary (like the stack). When the VM restarts, a new object with the bytes is created, you can fetch it via `supervisor.persisted_bytes`. If you want to throw away what you persisted, you can do `supervisor.persisted_bytes = None`.

Reply:
As @neradoc pointed `alarm.sleep_memory` has the same functionality. 
The main differences appear to be:
1. It is port dependent and seems to only be supported on esp32s2 (stably that is) 
2. It is grouped with the alarm module meaning implementing it is paired to the alarm module
3. It is fixed length where this __could__ be dynamic length 

## Possible Solution
merge alarm.sleep_memory and capsule as a `supervisor.saved_memory*` singleton with a max length that is included in only full builds of python.

For boards that do not have specialized features, a common implementation could do the on-stack (or safe place) swap. Here having a max length means you can't request to save something so large it would overflow the stack.

For ports that have specialized peripherals/memories, it is implemented on top of those.

\*still working title

Reply:
- Why limit the max length?
- Why take ram away from devices whether they use the feature or not?
- If you prefer, the previous 2 can be summed up as “why does this use static allocation?”
- `supervisor.persisted_bytes` was proposed as a name to use. I +1 that name.

> @WarriorOfWire: Agreed 1k is not a reasonable amount of memory for non-full builds of cp.

This is not what I attempted to communicate - I intended to represent that statically allocating a buffer that will be rarely used and is not even in a hot path is hostile to users’ memory. It is hostile whether it is a full build or not. 128 bits is 10 vectorio circles with xy locations. It’s not insignificant on any build and this memory cost ought only affect users of the api.

Choosing different buffer sizes per board also makes it more painful to use - and to Dan’s point which I share, it’s needless right?

Reply:
Thank you for re-framing your point and elaborating on Dan's. I now see how I miss interpreted both of your statements and guidance.
I hear you regarding statically allocating and how having the persisted data be any size tailors any cost involved not to the board or build but rather to the specific use case/call (?). That went right over my head for a bit: thank you for explaining.

I have a few questions to make sure I head in the right direction:
- Would it make sense for the message saving (whether by marking uncollectable or copying) would go somewhere in or between stop_mp, free_memory, or supervisor_move_memory?  [link](https://github.com/adafruit/circuitpython/blob/9a3f04a1b838988264a68c5d3c0e083a0c5c3163/main.c#L235-L237) 
- If so, where would you suggest putting the message restore? I'd hazard a guess it would go somewhere in or between stack_resize, allocate_remaining_memory, and start_mp? [link](https://github.com/adafruit/circuitpython/blob/9a3f04a1b838988264a68c5d3c0e083a0c5c3163/main.c#L293-L296)

Reply:
"the right direction" being towards `supervisor. persisted_bytes ` as Dan has suggested above.

Reply:
We would store `.persistent_bytes` in the `supervisor` module dictionary, which we would make mutable. We would treat it similarly to the `.wake_alarm` attribute in `alarm`, whose module dictionary is also mutable. `gc_collect_ptr()` is called specially on the `.wake_alarm` object, for instance.
https://github.com/adafruit/circuitpython/blob/9a3f04a1b838988264a68c5d3c0e083a0c5c3163/shared-bindings/alarm/__init__.c#L204
I do not remember if anything is done with the heap when the VM is stopped. That would have to be researched. If the heap is left alone, we don't need to copy the bytes until we start up the VM again, before we reinitialize the heap. The bytes could be copied onto the stack, the heap would be initialized, and then a `bytes` object would be allocated for the saved bytes and they would be copied in. Since the original object might be anywhere in heap storage, I don't think we can rely on it not being smashed up in some way when the heap is initialized (but I don't know the details of heap initialization either).

This is somewhat complicated and will require tracing through some internals stuff to make sure. I cannot say for sure "put some new code here and there" without re-understanding the detail VM and heap shutdown and startup code.

Reply:
I think others are better informed about the lifecycle of the Circuitpython init/teardown than me, though I do need an excuse to use this Segger. I’d love to see what Dan has to say there ❤️ 
Also to be clear, I don’t particularly care whether it’s a copy or marked objects. My care is that users should only “pay for what they use” as far as ram goes and just about any dynamic allocation strategy will accomplish that.